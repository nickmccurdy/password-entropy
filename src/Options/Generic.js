import {
  Checkbox,
  FormControlLabel,
  FormGroup,
  Typography
} from '@material-ui/core'
import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

export const examples = {
  'Lowercase Letters': 'a-z',
  'Uppercase Letters': 'A-Z',
  Numbers: '0-9',
  Symbols: '!@#$%^&*'
}

export default function Generic({ onGeneric, toggles }) {
  return (
    <FormGroup>
      {Object.entries(examples).map(([name, example]) => (
        <FormControlLabel
          key={name}
          control={
            <Checkbox
              name={name}
              checked={toggles[name]}
              onChange={onGeneric}
            />
          }
          label={
            <Fragment>
              {name}

              <Typography variant="caption" gutterBottom>
                {example}
              </Typography>
            </Fragment>
          }
        />
      ))}
    </FormGroup>
  )
}

Generic.propTypes = {
  onGeneric: PropTypes.func.isRequired,
  toggles: PropTypes.shape({
    'Lowercase Letters': PropTypes.bool.isRequired,
    'Uppercase Letters': PropTypes.bool.isRequired,
    Numbers: PropTypes.bool.isRequired,
    Symbols: PropTypes.bool.isRequired
  }).isRequired
}
