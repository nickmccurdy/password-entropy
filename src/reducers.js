import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
