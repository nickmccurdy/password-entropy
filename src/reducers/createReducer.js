import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
