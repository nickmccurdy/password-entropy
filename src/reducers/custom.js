import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
