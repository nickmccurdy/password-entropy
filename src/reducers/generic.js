import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
