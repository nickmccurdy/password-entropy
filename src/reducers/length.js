import pick from 'ramda/src/pick'

import createReducer from './createReducer'

export default createReducer('SET_LENGTH', 12)

export function setLength({ target: { value } }) {
  return {
    type: 'SET_LENGTH',
    payload: parseInt(value, 10)
  }
}

export const selector = pick(['length'])
