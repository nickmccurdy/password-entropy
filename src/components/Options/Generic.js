import startCase from 'lodash/startCase'
import entries from 'object.entries'
import PropTypes from 'prop-types'
import { always, map } from 'ramda'
import React, { Fragment } from 'react'
import { connect } from 'react-redux'
import { Input, FormGroup, Label } from 'reactstrap'

import { selector, toggleGeneric, toggles } from '../../reducers/generic'

export function Generic(props) {
  return (
    <Fragment>
      {entries(toggles).map(([name, { example }]) => (
        <FormGroup key={name} check>
          <Label check>
            <Input type="checkbox" name={name} checked={props[name]} onChange={props.toggleGeneric} />{' '}
            {startCase(name)} <small>({example})</small>
          </Label>
        </FormGroup>
      ))}
    </Fragment>
  )
}

Generic.propTypes = {
  ...map(always(PropTypes.bool.isRequired), toggles),
  toggleGeneric: PropTypes.func.isRequired
}

export default connect(selector, { toggleGeneric })(Generic)
