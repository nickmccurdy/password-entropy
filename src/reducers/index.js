import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
