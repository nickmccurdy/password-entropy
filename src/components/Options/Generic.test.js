import { shallow } from 'enzyme'
import React from 'react'

import { Generic } from './Generic'
import { defaultToggles } from '../../reducers/generic'

test('Generic', () => {
  expect(shallow(<Generic toggleGeneric={() => {}} {...defaultToggles} />)).toMatchSnapshot()
})
