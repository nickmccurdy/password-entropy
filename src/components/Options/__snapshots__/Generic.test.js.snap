// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic 1`] = `
<Fragment>
  <FormGroup
    check={true}
    key="Letters"
    tag="div"
  >
    <Label
      check={true}
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Input
        checked={true}
        name="Letters"
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        type="checkbox"
      />
      Letters
       
      <small>
        (
        a-z
        )
      </small>
    </Label>
  </FormGroup>
  <FormGroup
    check={true}
    key="Capital Letters"
    tag="div"
  >
    <Label
      check={true}
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Input
        checked={true}
        name="Capital Letters"
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        type="checkbox"
      />
      Capital Letters
       
      <small>
        (
        A-Z
        )
      </small>
    </Label>
  </FormGroup>
  <FormGroup
    check={true}
    key="Numbers"
    tag="div"
  >
    <Label
      check={true}
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Input
        checked={true}
        name="Numbers"
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        type="checkbox"
      />
      Numbers
       
      <small>
        (
        0-9
        )
      </small>
    </Label>
  </FormGroup>
  <FormGroup
    check={true}
    key="Symbols"
    tag="div"
  >
    <Label
      check={true}
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Input
        checked={true}
        name="Symbols"
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        type="checkbox"
      />
      Symbols
       
      <small>
        (
        !@#$%^&*
        )
      </small>
    </Label>
  </FormGroup>
</Fragment>
`;
